### PR TITLE
Foursquare Venue Widget

### PR DESCRIPTION
**Pull request by tobiasheine**
### Scope of the PR

Adding a dashboard widget which displays the `mayor` of a given `foursquare venue`.

### Details

The first iteration uses the `classic` design. A follow-up task will be to create a custom design containing the profile image and some more information.

### Screenshots

| After |
| --- |
| screen shot 2017-01-05 at 18 18 05 |

Paired with @florianmski

### CODE REVIEW

File: node-server/server/widget/foursquare-venue.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  _getMayor() {
+    const REQUEST = {
+      url: this._buildUrl(),
+      gzip: true
+    };
+    return this.httpClient(REQUEST).then(this._parseResponse.bind(this));
+  }
+
+  _buildUrl() {
+    return BASE_URL + `${this.venueId}?v=20170101&client_id=${this.clientId}&client_secret=${this.clientSecret}&m=swarm`;
```
Comment (frapontillo) [inline node-server/server/widget/foursquare-venue.js]: What is the `v` parameter?

File: node-server/server/widget/foursquare-venue.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  _parseResponse(body) {
+    const jsonBody = JSON.parse(body);
+    const venue = jsonBody.response.venue;
+    if (venue) {
+      if (venue.mayor) {
+        return Promise.resolve({ venue: venue, mayor: venue.mayor });
+      } else {
```
Comment (frapontillo) [inline node-server/server/widget/foursquare-venue.js]: You ~can~ should get rid of the `else` path and just `return`.

File: node-server/server/widget/foursquare-venue.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return Promise.reject(`No mayor for venue "${venue.name}"`);
+      }
+    } else {
```
Comment (frapontillo) [inline node-server/server/widget/foursquare-venue.js]: Same here.

**Comment by mr-archano** [comment]
_me likes_ 👍 (can't tell about the actual code tho)

File: node-server/server/widget/foursquare-venue.spec.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+describe('FoursquareVenue', () => {
+
+  let foursquareVenue, mockHttpClient;
+
+  beforeEach(function() {
+    mockHttpClient = new MockHttpClient();
+  });
+
+  it('should return error payload when no venueId defined', (done) => {
+    foursquareVenue = new FoursquareVenue({ foursquare: {clientId: 'foo', clientSecret: 'bar'}}, {}, mockHttpClient.request());
```
Comment (frapontillo) [inline node-server/server/widget/foursquare-venue.spec.js]: This is called for every test, you can move it in the `beforeEach`.

File: node-server/server/configs/sample.json [hunk 5]
```diff
@@ -45,5 +45,6 @@
     "eventName": "Christmas",
     "eventDate": "2016-12-25",
     "template": "classicXmas"
-  }
+  },
+  "foursquare-venue": {"venueId": "56f0fe62cd107047e6c74bac"}
```
Comment (eduardb) [inline node-server/server/configs/sample.json]: Isn't it possible to have a more user-friendly way to identify the venue? Like name/address for instance?

Comment (florianmski) [inline node-server/server/configs/sample.json]: Please format as the rest of the file

File: node-server/server/widget/foursquare-venue.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    if (!this.clientId) {
+      return this._errorRule('No "clientId" defined in the global configuration');
+    }
+
+    if (!this.clientSecret) {
+      return this._errorRule('No "clientSecret" defined in the global configuration');
+    }
+
```
Comment (florianmski) [inline node-server/server/widget/foursquare-venue.js]: Not sure about all these white spaces

File: node-server/server/widget/foursquare-venue.js (hunk 1 above)
Comment (florianmski) [inline node-server/server/widget/foursquare-venue.js]: That's the version of the api that will be chose based on the date if I recall correctly, @tobiasheine ?

File: node-server/server/widget/foursquare-venue.spec.js (hunk 4 above)
Comment (florianmski) [inline node-server/server/widget/foursquare-venue.spec.js]: Nope, we need to pass different arguments to test we fail correctly when they are not there

File: node-server/server/widget/foursquare-venue.spec.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    mockHttpClient.enqueue(`
+      {
+      	"meta": {},
+      	"response": {
+      		"venue": {
+      			"id": "56f0fe62cd107047e6c74bac",
+      			"name": "Novoda"
+      		}
+      	}
+      }
```
Comment (florianmski) [inline node-server/server/widget/foursquare-venue.spec.js]: Formatting is off throughout the file for the response enqueued

File: node-server/server/widget/foursquare-venue.js (hunk 1 above)
Comment (tobiasheine) [inline node-server/server/widget/foursquare-venue.js]: exactly, will extract a constant to make it more obvious

File: node-server/server/widget/foursquare-venue.spec.js (hunk 7 above)
Comment (tobiasheine) [inline node-server/server/widget/foursquare-venue.spec.js]: is there a way to autoformat? I remember you told me sth. about a plugin I should install...

File: node-server/server/configs/sample.json (hunk 5 above)
Comment (tobiasheine) [inline node-server/server/configs/sample.json]: Unfortunately we're limited by the [venue endpoint](https://developer.foursquare.com/docs/venues/search). 
An alternative would be using the [search endpoint](https://developer.foursquare.com/docs/venues/search) but then we could not be sure that the result contains the venue we're looking for. Also you can change the name/location for a venue after it is created.

File: node-server/server/widget/foursquare-venue.spec.js (hunk 7 above)
Comment (florianmski) [inline node-server/server/widget/foursquare-venue.spec.js]: @tobiasheine have a look at the description of #38

**Comment by florianmski** [comment]
Leaving it to @eduardb for the merge 👍

File: node-server/server/configs/sample.json (hunk 5 above)
Comment (tasomaniac) [inline node-server/server/configs/sample.json]: For all kind of widgets, we can support some setup functionality. Maybe we can have add-widget page where you select the FSq widget than use the search endpoint to find the venue. Then selecting the venue would persist the id somewhere etc.

**Comment by tasomaniac** [comment]
As far as I understand, our dashboards is not office specific right? All the screens are running the same. Which has the advantage because we are seeing the same thing at a given time. How are we planning to have different venues for each office?

**Comment by florianmski** [comment]
@tasomaniac no, we can have different dashboards so we can be office / room / project specific. If you have more questions on our current setup let's do that IRL as this is an OSS project so let's try to keep a more generic discussion :)

File: node-server/server/configs/sample.json (hunk 5 above)
Comment (florianmski) [inline node-server/server/configs/sample.json]: @tasomaniac yeah I agree that would be ideal but will require quite some work

**Pull request merged**